File: Graph/Python/Number of Islands.py
# Problem link: https://leetcode.com/problems/number-of-islands/description/

# Approach: 1 (DFS)
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        visit = set()
        result, rows, cols = 0, len(grid), len(grid[0])
        directions = [[0, 1], [1, 0], [-1, 0], [0, -1]]

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1' and (i, j) not in visit:
                    self.dfs(i, j, grid, visit, directions)
                    result += 1
        
        return result
    
    def dfs(self, row, col, grid, visit, directions):
        if row < 0 or col < 0 or row == len(grid) or col == len(grid[0]) or grid[row][col] == '0' or (row, col) in visit:
            return
        
        visit.add((row, col))

        for r, c in directions:
            self.dfs(row + r, col + c, grid, visit, directions)  

# Approach: 2 (BFS)
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        rows, cols, result = len(grid), len(grid[0]), 0
        visit, directions = set(), [[1, 0], [0, 1], [-1, 0], [0, -1]]

        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1' and (i, j) not in visit:
                    self.bfs(i, j, visit, grid, directions)
                    result += 1

        return result
        
    
    def bfs(self, row, col, visit, grid, directions):
        queue = deque([(row, col)])
        rows, cols = len(grid), len(grid[0])

        while queue:
            row, col = queue.popleft()

            if (row, col) in visit:
                continue
            
            visit.add((row, col))

            for r, c in directions:
                if r + row >= 0 and r + row < rows and c + col >= 0 and c + col < cols and grid[r + row][c + col] == '1':
                    queue.append((r + row, c + col))

File: Graph/Java/Number of Islands.java
// Approach: 1 (DFS)
// Time: O(n * m)
// Space: o(n * m)

class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = 0;
        boolean[][] visited = new boolean[rows][cols];
        int[][] directions = {{-1, 0}, {0, -1}, {1, 0}, {0, 1}};

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == '1' && !visited[i][j]) {
                    result++;
                    this.dfs(i, j, visited, grid, directions);
                }
            }
        }

        return result;
    }

    private void dfs(int row, int col, boolean[][] visited, char[][] grid, int[][] directions) {
        if(row < 0 || col < 0 || row == grid.length || col == grid[0].length || grid[row][col] == '0' || visited[row][col])
            return;
        
        visited[row][col] = true;

        for(int[] rc : directions)
            this.dfs(row + rc[0], col + rc[1], visited, grid, directions);
    }
}



// Approach: 2 (BFS)
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int numIslands(char[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = 0;
        boolean[][] visit = new boolean[rows][cols];

        for(int i = 0; i < rows; i++) {
            for(int j = 0; j < cols; j++) {
                if(grid[i][j] == '1' && !visit[i][j]) {
                    result++;
                    this.bfs(grid, i, j, visit);
                }
            }
        }

        return result;
    }

    private void bfs(char[][] grid, int row, int col, boolean[][] visit) {
        Queue<int[]> queue = new LinkedList<>();
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
        int rows = grid.length, cols = grid[0].length;
        queue.offer(new int[] {row, col});

        while(!queue.isEmpty()) {
            int[] store = queue.poll();

            if(visit[store[0]][store[1]])
                continue;
            
            visit[store[0]][store[1]] = true;

            for(int[] rc : directions) {
                if(rc[0] + store[0] >= 0 && rc[0] + store[0] < rows && rc[1] + store[1] >= 0 && rc[1] + store[1] < cols)
                    if(grid[rc[0] + store[0]][rc[1] + store[1]] != '0')
                        queue.offer(new int[] {store[0] + rc[0], store[1] + rc[1]});
            }
        }
    }
}
